lib/config.js

```diff
@@ -212,127 +87,276 @@ class Config {
+            projectConfigPath = ConfigFile.getFilenameForDirectory(this.options.cwd),
+            searched = [],
+            configs = [];
+        let rootPath,
+            cache;
+
+        for (const localConfigFile of localConfigFiles) {
+            const localConfigDirectory = path.dirname(localConfigFile);
+
+            cache = configCache.getHierarchyLocalConfigs(localConfigDirectory);
+            if (cache) {
+                break;
+            }
 
-            config = ConfigOps.merge(config, this.useSpecificConfig);
-        }
+            // Don't consider the personal config file in the home directory,
+            // except if the home directory is the same as the current working directory
+            if (localConfigDirectory === PERSONAL_CONFIG_DIR && localConfigFile !== projectConfigPath) {
+                continue;
+            }
 
-        // Step 5: Merge in command line environments
-        debug("Merging command line environment settings");
-        config = ConfigOps.merge(config, { env: this.env });
+            // If root flag is set, don't consider file if it is above root
+            if (rootPath && !pathIsInside(path.dirname(localConfigFile), rootPath)) {
+                continue;
+            }
 
-        // Step 6: Merge in command line rules
-        if (this.options.rules) {
-            debug("Merging command line rules");
-            config = ConfigOps.merge(config, { rules: this.options.rules });
-        }
+            debug(`Loading ${localConfigFile}`);
+            const localConfig = ConfigFile.loadCached(localConfigFile);
 
-        // Step 7: Merge in command line globals
-        config = ConfigOps.merge(config, { globals: this.globals });
+            // Ignore empty config files
+            if (!localConfig) {
+                continue;
+            }
 
-        // Only override parser if it is passed explicitly through the command line or if it's not
-        // defined yet (because the final object will at least have the parser key)
-        if (this.parser || !config.parser) {
-            config = ConfigOps.merge(config, {
-                parser: this.parser
-            });
-        }
+            // Check for root flag
+            if (localConfig.root === true) {
+                rootPath = path.dirname(localConfigFile);
+            }
 
-        if (this.parserOptions) {
-            config = ConfigOps.merge(config, {
-                parserOptions: this.parserOptions
-            });
+            debug(`Using ${localConfigFile}`);
+            configs.push(localConfig);
+            searched.push(localConfigDirectory);
         }
 
-        // Step 8: Merge in command line plugins
-        if (this.options.plugins) {
-            debug("Merging command line plugins");
-            Plugins.loadAll(this.options.plugins);
-            config = ConfigOps.merge(config, { plugins: this.options.plugins });
-        }
+        if (!configs.length && !cache && !this.useSpecificConfig) {
```

Comment:
This logic is hard to follow. Would it be possible to split it into multiple functions?

Specifically, it's not clear to me how the presence of a cache changes the control flow. Intuitively, I would expect the cache to be an alternate way to populate `configs`, but in that case its effects should be isolated, and it shouldn't be necessary to refer to the cache separately afterwards.